pkg/controller/acmechallenges/sync.go

```diff
@@ -72,6 +74,19 @@ func (c *controller) Sync(ctx context.Context, ch *cmacme.Challenge) (err error)
 		return c.handleFinalizer(ctx, ch)
 	}
 
+	// if dns01-check-recreate-period is enabled and the pending challenge's creation time
+	// is more than DNS01CheckRecreatePeriod ealier from now delete the stale challenge so it can be recreated
+	rp := c.DNS01CheckRecreatePeriod
+	if rp > 0 && ch.Spec.Type == v1.ACMEChallengeTypeDNS01 && ch.Status.State == v1.Pending &&
+		ch.CreationTimestamp.Add(rp).Before(time.Now()) {
+		if err := c.cmClient.AcmeV1().Challenges(ch.Namespace).Delete(ctx, ch.Name, metav1.DeleteOptions{}); err != nil {
```

Comment:
I've not had a chance to fully review this, but deleting the object we are currently syncing here seems a bit odd 👀

Could this have been alternatively implemented by having the acmechallenges controller call Present again, instead of deleting the resource?

As an example, if someone were to be using the Challenge API directly (instead of via a Certificate resource), this patch wouldn't really work for them as you're relying on the caller to recreate the resource